scilpy/tracking/cmc_tracking.cl

```diff
@@ -0,0 +1,492 @@
+              const size_t n_seeds, const float max_cos_theta,
+              __global const float* sh_coeffs,
+              __global const float* sf_max,
+              __global const float* sf_rand_values,
+              __global const float* vertices,
+              __global const float* sh_to_sf_mat,
+              __global const float* map_include,
+              __global const float* map_exclude,
+              __global const float* continue_rand_values,
+              __global const float* include_rand_values,
+              __global float* out_streamlines)
+{
+    const int max_length = is_forward ? MAX_LENGTH / 2 :
+                           current_length + MAX_LENGTH / 2;
+    float continue_randv =  continue_rand_values[get_flat_index(seed_indice, current_length,
+                                                                0, 0, n_seeds, MAX_LENGTH, 1)];
+    bool can_propagate = propagation_can_continue(map_include, map_exclude,
+                                                  continue_randv, last_pos);
+
+    while(can_propagate)
+    {
+        if(!(current_length < max_length))
+        {
+            // when the propagation can continue, but maximum track length is
+            // reached, we return the streamline as is.
+            return current_length;
+        }
+        // Sample SF at position.
+        // Get SH at position.
+        float odf_sh[IM_N_COEFFS];
+        get_value(sh_coeffs, IM_N_COEFFS, last_pos, SH_INTERP_NN, odf_sh);
+
+        // Convert SH to SF.
+        const float curr_sf_max = sf_max[get_flat_index(last_pos.x, last_pos.y,
+                                                        last_pos.z, 0, IM_X_DIM,
+                                                        IM_Y_DIM, IM_Z_DIM)];
+        float odf_sf[N_DIRS];
+        sh_to_sf(odf_sh, sh_to_sf_mat, curr_sf_max, current_length == 1,
+                 vertices, last_dir, max_cos_theta, odf_sf);
+
+        // Sample distribution.
+        const float randv = sf_rand_values[get_flat_index(seed_indice, current_length,
+                                                          0, 0, n_seeds, MAX_LENGTH, 1)];
+        const int vert_indice = sample_sf(odf_sf, randv);
+
+        if(vert_indice >= 0)
+        {
+            const float3 direction = {
+                vertices[get_flat_index(vert_indice, 0, 0, 0, N_DIRS, 3, 1)],
+                vertices[get_flat_index(vert_indice, 1, 0, 0, N_DIRS, 3, 1)],
+                vertices[get_flat_index(vert_indice, 2, 0, 0, N_DIRS, 3, 1)]
+            };
+
+            // Try step.
+            const float3 next_pos = last_pos + STEP_SIZE * direction;
+            last_dir = normalize(next_pos - last_pos);
+            last_pos = next_pos;
+        }
+        else
+        {
+            // NOTE: reached a region where we can't track; exclude streamline
+            // ONLY MAKES SENSE IN THE SPECIFIC CASE OF SHORT-TRACKS TRACKING.
+            return 0;
```

Comment:
Here, when there is no direction available inside the tracking cone for a given step, the tracking immediately rejects the streamline. However in the PFT paper, the tracking should keep the previous directions for some `delta_undeviated` steps (if I'm not mistaken).